File: spell_recognizer_2/save/spell_matcher.py
"""
Spell Matching -- Relational Layer on top of QRecognizer
==========================================================
$Q (recognizer.py) intentionally throws away position: every gesture is
uniformly scaled into a fixed frame and re-centered on its own centroid
before scoring (see QRecognizer._preprocess), because classifying "is this
stroke a triangle" has to work no matter where on the canvas it was drawn.

A *spell*, though, is defined by exactly the information $Q throws away:
not just "there's a star and four triangles here", but "the triangles sit
roughly north/east/south/west of the star, at roughly this distance from
it". That's a relational/layout question, not a shape-classification
question, so it lives in its own layer instead of being bolted onto
QRecognizer:

  Layer 1 (recognizer.py, unchanged): classify each physically-separate
    stroke cluster into a named, leveled SceneFeature (circle, triangle, star_5,
    ...) via QRecognizer.recognize_scene. Position is deliberately ignored
    here.

  Layer 2 (this module): given the List[SceneFeature] recognize_scene just
    produced -- which still carries each SceneFeature's RAW, unnormalized scene
    points (SceneFeature.points is the pre-_preprocess data) -- work out
    where each SceneFeature sits *relative to the whole drawing* and match that
    layout against a library of spell definitions.

Every SceneFeature in a spell (including circles/lines/rings -- there's no
special "center" node type; everything is a SceneFeature, some just happen to
sit at distance ~0) is described the same way: a shape name, a distance
from the spell's own center (normalized to the drawing's own size, so
spells are scale-invariant regardless of how big they're drawn), and an
optional angle sector ("between compass bearing X and Y") for SceneFeatures
whose *position* -- not just presence -- matters, like "a triangle north of
center". Compass angle, not math angle: 0=N, 90=E, 180=S, 270=W, matching
how someone points at a compass rather than how atan2 returns radians.

What this module deliberately does NOT give you: rotation invariance. The
spell center is derived from the drawing itself, but "north" is always
"up on the page" -- a correctly-drawn spell rotated 30 degrees will not
match. That's a real tradeoff (see module-level note at the bottom for the
one-line change that would trade it back for angle-relative-to-an-anchor-
SceneFeature instead of true north), made deliberately because "triangle to the
north" is meaningless once you drop absolute orientation.

Layout of this file:
  1. Geometry helpers        -- compass_angle, angle_in_range,
                                 compute_positions (SceneFeature -> relative
                                 distance/angle)
  2. Spell definitions        -- SpellSceneFeatureSlot, SpellDefinition (pure
                                  data -- persistence lives in spell_store.py)
  3. Matching                 -- SpellMatchResult, match_spell,
                                  match_best_spell
"""

import math
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple, TYPE_CHECKING

from merge_intersecting_strokes import Point

if TYPE_CHECKING:
    from recognizer import SceneFeature

from config import (
    DEFAULT_SPELL_DIST_TOLERANCE,
    DEFAULT_SPELL_ANGLE_TOLERANCE,
    DEFAULT_SPELL_MIN_SCORE,
    SPELL_CENTER_EPSILON,
)


# =============================================================================
# 1. Geometry helpers
# =============================================================================

def compass_angle(dx: float, dy: float) -> float:
    """
    Bearing of the vector (dx, dy) in COMPASS convention: 0=North, 90=East,
    180=South, 270=West, increasing clockwise -- not math convention (0 on
    the positive x-axis, counter-clockwise), because spell authors think in
    "north of center", not "positive x-axis". Screen y grows downward, so
    "up" is -dy; that's the only reason dy is negated here.
    """
    return math.degrees(math.atan2(dx, -dy)) % 360.0


def angle_in_range(angle: float, min_angle: float, max_angle: float) -> bool:
    """
    True if `angle` (compass bearing, any real number) falls inside the
    sector [min_angle, max_angle] going CLOCKWISE from min_angle to
    max_angle. Handles the wraparound case (e.g. min=315, max=45 spans
    through the 360/0 seam) by checking whether the sector itself wraps,
    rather than assuming min_angle <= max_angle the way a plain range check
    would.
    """
    a = angle % 360.0
    lo = min_angle % 360.0
    hi = max_angle % 360.0
    if lo <= hi:
        return lo <= a <= hi
    return a >= lo or a <= hi


def _sector_overshoot(angle: float, min_angle: float, max_angle: float) -> float:
    """
    For an `angle` OUTSIDE the [min_angle, max_angle] sector, how many
    degrees past the nearer edge it sits. Used only for soft scoring (see
    _slot_score) -- angle_in_range is the hard gate; this is what turns a
    near-miss into a partial-credit score instead of a flat zero.
    """
    lo, hi = min_angle % 360.0, max_angle % 360.0
    span = (hi - lo) % 360.0
    rel = (angle % 360.0 - lo) % 360.0  # position within the sector's own frame, 0..360
    # rel <= span would mean "inside", which the caller already ruled out;
    # the distance past the far (hi) edge going forward vs. past the near
    # (lo) edge going backward -- take whichever is smaller.
    return min(rel - span, 360.0 - rel)


def _centroid_xy(points: List[Point]) -> Tuple[float, float]:
    x = sum(p.x for p in points) / len(points)
    y = sum(p.y for p in points) / len(points)
    return x, y


def _bounding_diagonal(all_points: List[Point]) -> float:
    """Diagonal of the bounding box over every point in the whole drawing --
    what SceneFeature distances are normalized against, so a spell drawn big or
    small still matches the same normalized distances."""
    xs = [p.x for p in all_points]
    ys = [p.y for p in all_points]
    w = max(xs) - min(xs)
    h = max(ys) - min(ys)
    return math.hypot(w, h)


@dataclass
class PositionedSceneFeature:
    """
    A SceneFeature re-expressed relative to the whole drawing: everything
    the matcher actually needs, decoupled from recognizer.py's SceneFeature
    type so this module only depends on it for type-checking, not at
    runtime (see the TYPE_CHECKING import above).
    """
    shape: str
    distance: float           # normalized: 0.0 = at the spell's own center
    angle: float               # compass bearing, degrees
    source: "SceneFeature"     # original SceneFeature, kept for overlays/debugging


def compute_positions(SceneFeatures: List["SceneFeature"]) -> List[PositionedSceneFeature]:
    """
    Turns a flat List[SceneFeature] (as returned by
    QRecognizer.recognize_scene) into PositionedSceneFeatures: each SceneFeature's
    shape name plus its distance/angle relative to the drawing's OWN center
    (the centroid of every SceneFeature's own centroid, each SceneFeature weighted
    equally regardless of how many points/strokes made it up -- so a
    fiddly, many-point star doesn't drag the center toward itself more
    than a simple two-point line would) and the drawing's OWN size
    (bounding-box diagonal over every point in the whole drawing).

    Only SceneFeatures with a real recognized name are usable for spell
    matching (an unrecognized/rejected cluster can't fill a spell slot), so
    those are silently skipped here rather than raising -- a spell match
    against a scene with some unrecognized clutter in it should still be
    possible.
    """
    named = [f for f in SceneFeatures if f.result.name is not None]
    if not named:
        return []

    centroids = [_centroid_xy(f.points) for f in named]
    center_x = sum(c[0] for c in centroids) / len(centroids)
    center_y = sum(c[1] for c in centroids) / len(centroids)

    all_points = [p for f in named for p in f.points]
    diag = _bounding_diagonal(all_points)
    if diag <= 1e-9:
        diag = 1.0  # degenerate single-point drawing -- avoid div-by-zero

    positioned = []
    for feat, (cx, cy) in zip(named, centroids):
        dx, dy = cx - center_x, cy - center_y
        raw_dist = math.hypot(dx, dy)
        positioned.append(PositionedSceneFeature(
            shape=feat.result.name,
            distance=raw_dist / diag,
            angle=compass_angle(dx, dy),
            source=feat,
        ))
    return positioned


# =============================================================================
# 2. Spell definitions (pure data; see spell_store.py for JSON persistence)
# =============================================================================

@dataclass
class SpellSceneFeatureSlot:
    """
    One required SceneFeature within a spell, described relationally rather
    than by absolute position -- see this module's docstring. `id` only
    needs to be unique within its own SpellDefinition (it's how a match
    result reports which scene SceneFeature filled which slot).
    """
    id: int
    shape: str
    distance: float
    tolerance_dist: float = DEFAULT_SPELL_DIST_TOLERANCE
    min_angle: Optional[float] = None   # None -> angle unconstrained (any position OK)
    max_angle: Optional[float] = None
    tolerance_angle: float = DEFAULT_SPELL_ANGLE_TOLERANCE  # soft margin for scoring just outside the sector

    def angle_constrained(self) -> bool:
        return self.min_angle is not None and self.max_angle is not None


@dataclass
class SpellDefinition:
    name: str
    SceneFeatures: List[SpellSceneFeatureSlot]
    min_score: float = DEFAULT_SPELL_MIN_SCORE


# =============================================================================
# 3. Matching
# =============================================================================

@dataclass
class SpellMatchResult:
    name: Optional[str]
    score: float
    accepted: bool
    # slot.id -> matched PositionedSceneFeature, for whichever slots got filled;
    # a slot missing from this dict means nothing in the scene satisfied it
    # well enough to be worth assigning (useful for "you're missing the
    # north triangle" style feedback even on a rejected match).
    assignment: Dict[int, PositionedSceneFeature] = field(default_factory=dict)


def _slot_score(slot: SpellSceneFeatureSlot, feat: PositionedSceneFeature) -> float:
    """
    Score in [0.0, 1.0] for matching `feat` to `slot`, given they already
    share the same shape (shape match is a hard gate, checked by the
    caller via the by-shape candidate grouping in match_spell -- a
    triangle can never fill a circle's slot no matter how well-positioned
    it is). Blends a distance-closeness score with an angle-closeness
    score (the latter only when the slot actually constrains angle),
    rather than being a hard pass/fail, so near-misses still produce a
    meaningful score for feedback instead of just "no match" -- the same
    philosophy _cloud_distance/_blend_avg_max use in recognizer.py.
    """
    dist_diff = abs(feat.distance - slot.distance)
    if dist_diff > slot.tolerance_dist:
        return 0.0
    dist_score = 1.0 - (dist_diff / slot.tolerance_dist if slot.tolerance_dist > 1e-9 else 0.0)

    if not slot.angle_constrained() or slot.distance <= SPELL_CENTER_EPSILON:
        return dist_score  # angle is meaningless this close to center, or the slot doesn't care

    if angle_in_range(feat.angle, slot.min_angle, slot.max_angle):
        return dist_score

    overshoot = _sector_overshoot(feat.angle, slot.min_angle, slot.max_angle)
    angle_score = max(0.0, 1.0 - overshoot / max(slot.tolerance_angle, 1e-9))
    if angle_score <= 0.0:
        return 0.0
    return dist_score * angle_score


def match_spell(spell: SpellDefinition, scene_SceneFeatures: List["SceneFeature"]
                 ) -> SpellMatchResult:
    """
    Finds the best injective assignment of `scene_SceneFeatures` to `spell`'s
    slots (each scene SceneFeature can fill at most one slot) via backtracking:
    for every slot, only shape-matching, positive-score candidates are
    ever tried, plus a "leave unfilled" branch. Spell templates are small
    (a handful of SceneFeatures), so this is cheap -- no need for a general
    subgraph-isomorphism solver.

    Score = mean _slot_score over every slot in the spell (an unfilled
    slot contributes 0 to that average). `accepted` requires BOTH every
    slot to be filled (score alone can't distinguish "filled badly" from
    "empty" -- both drag a small spell's average down similarly) AND the
    resulting score to clear `spell.min_score`.
    """
    positioned = compute_positions(scene_SceneFeatures)

    # Group candidates by shape up front so each slot only ever considers
    # scene SceneFeatures that could possibly fill it.
    by_shape: Dict[str, List[int]] = {}
    for idx, feat in enumerate(positioned):
        by_shape.setdefault(feat.shape, []).append(idx)

    slots = spell.SceneFeatures
    used = [False] * len(positioned)
    current: Dict[int, int] = {}

    best: Dict[int, int] = {}
    best_score = -1.0

    def backtrack(slot_pos: int, running_score: float) -> None:
        nonlocal best, best_score
        if slot_pos == len(slots):
            avg = running_score / len(slots) if slots else 0.0
            if avg > best_score:
                best_score = avg
                best = dict(current)
            return

        slot = slots[slot_pos]
        for idx in by_shape.get(slot.shape, []):
            if used[idx]:
                continue
            score = _slot_score(slot, positioned[idx])
            if score <= 0.0:
                continue
            used[idx] = True
            current[slot.id] = idx
            backtrack(slot_pos + 1, running_score + score)
            del current[slot.id]
            used[idx] = False

        # Also try leaving this slot unfilled, so a spell missing one
        # required SceneFeature still yields the best possible PARTIAL match
        # (useful for near-miss feedback) instead of no assignment at all.
        backtrack(slot_pos + 1, running_score)

    backtrack(0, 0.0)

    assignment = {slot_id: positioned[idx] for slot_id, idx in best.items()}
    all_filled = len(assignment) == len(slots)
    final_score = max(best_score, 0.0)
    accepted = all_filled and final_score >= spell.min_score

    return SpellMatchResult(
        name=spell.name if accepted else None,
        score=final_score,
        accepted=accepted,
        assignment=assignment,
    )


def match_best_spell(spells: List[SpellDefinition], scene_SceneFeatures: List["SceneFeature"]
                      ) -> Optional[SpellMatchResult]:
    """
    Tries every known spell against the same scene and returns whichever
    ACCEPTED result scored highest, or None if nothing accepted. Mirrors
    QRecognizer.recognize's "best template wins" shape, just one level up
    (best SPELL instead of best per-level template).
    """
    best_result: Optional[SpellMatchResult] = None
    for spell in spells:
        result = match_spell(spell, scene_SceneFeatures)
        if result.accepted and (best_result is None or result.score > best_result.score):
            best_result = result
    return best_result


# -----------------------------------------------------------------------------
# Note on rotation invariance (see module docstring)
# -----------------------------------------------------------------------------
# To match a spell drawn at any rotation, angles would need to be measured
# relative to one designated "anchor" SceneFeature's bearing from center instead
# of true north -- i.e. `angle = (compass_angle(dx, dy) - anchor_bearing) %
# 360` for every SceneFeature, anchor included (which then sits at angle 0 by
# definition). That's a change to compute_positions (needs to know which
# SceneFeature is the anchor) and to how SpellSceneFeatureSlot.min_angle/max_angle
# are interpreted (relative bearing from the anchor, not compass bearing) --
# deliberately NOT done here since it trades away the ability to express
# "must be drawn upright", which most spells probably want.

File: spell_recognizer_2/cpp_implementation/spell_matcher.cpp
// spell_matcher.cpp
#include "spell_matcher.hpp"

#include <algorithm>
#include <cmath>
#include <functional>
#include <limits>
#include <map>
#include <stdexcept>
#include <unordered_map>
#include <unordered_set>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

namespace qrec {

// =============================================================================
// 1. Geometry helpers
// =============================================================================

double compass_angle(double dx, double dy) {
    double deg = std::fmod(std::atan2(dx, -dy) * 180.0 / M_PI, 360.0);
    if (deg < 0) deg += 360.0;
    return deg;
}

namespace {

double norm360(double a) {
    double r = std::fmod(a, 360.0);
    if (r < 0) r += 360.0;
    return r;
}

}  // namespace

bool angle_in_range(double angle, double min_angle, double max_angle) {
    double a = norm360(angle);
    double lo = norm360(min_angle);
    double hi = norm360(max_angle);
    if (lo <= hi) {
        return lo <= a && a <= hi;
    }
    return a >= lo || a <= hi;
}

namespace {

// For an `angle` OUTSIDE the [min_angle, max_angle] sector, how many
// degrees past the nearer edge it sits. Used only for soft scoring --
// angle_in_range is the hard gate.
double sector_overshoot(double angle, double min_angle, double max_angle) {
    double lo = norm360(min_angle), hi = norm360(max_angle);
    double span = norm360(hi - lo);
    double rel = norm360(norm360(angle) - lo);
    return std::min(rel - span, 360.0 - rel);
}

std::pair<double, double> centroid_xy(const std::vector<Point>& points) {
    double x = 0.0, y = 0.0;
    for (const auto& p : points) {
        x += p.x;
        y += p.y;
    }
    x /= static_cast<double>(points.size());
    y /= static_cast<double>(points.size());
    return {x, y};
}

// Average distance of `points` from their own centroid (cx, cy).
double mean_radius(const std::vector<Point>& points, double cx, double cy) {
    double sum = 0.0;
    for (const auto& p : points) {
        sum += std::hypot(p.x - cx, p.y - cy);
    }
    return sum / static_cast<double>(points.size());
}

// Diagonal of the bounding box over every point in the whole drawing.
double bounding_diagonal(const std::vector<Point>& all_points) {
    double min_x = all_points[0].x, max_x = all_points[0].x;
    double min_y = all_points[0].y, max_y = all_points[0].y;
    for (const auto& p : all_points) {
        min_x = std::min(min_x, p.x);
        max_x = std::max(max_x, p.x);
        min_y = std::min(min_y, p.y);
        max_y = std::max(max_y, p.y);
    }
    double w = max_x - min_x;
    double h = max_y - min_y;
    return std::hypot(w, h);
}

}  // namespace

std::vector<PositionedFeature> compute_positions(const std::vector<std::shared_ptr<Feature>>& features) {
    std::vector<std::shared_ptr<Feature>> named;
    for (const auto& f : features) {
        if (f->result.name.has_value()) named.push_back(f);
    }
    if (named.empty()) return {};

    std::vector<std::pair<double, double>> centroids;
    centroids.reserve(named.size());
    for (const auto& f : named) centroids.push_back(centroid_xy(f->points));

    double center_x = 0.0, center_y = 0.0;
    for (const auto& c : centroids) {
        center_x += c.first;
        center_y += c.second;
    }
    center_x /= static_cast<double>(centroids.size());
    center_y /= static_cast<double>(centroids.size());

    std::vector<Point> all_points;
    for (const auto& f : named) all_points.insert(all_points.end(), f->points.begin(), f->points.end());
    double diag = bounding_diagonal(all_points);
    if (diag <= 1e-9) diag = 1.0;  // degenerate single-point drawing -- avoid div-by-zero

    std::vector<PositionedFeature> positioned;
    positioned.reserve(named.size());
    for (size_t i = 0; i < named.size(); ++i) {
        const auto& feat = named[i];
        double cx = centroids[i].first, cy = centroids[i].second;
        double dx = cx - center_x, dy = cy - center_y;
        double raw_dist = std::hypot(dx, dy);
        double own_radius = mean_radius(feat->points, cx, cy);

        PositionedFeature pf;
        pf.shape = *feat->result.name;
        pf.distance = raw_dist / diag;
        pf.angle = compass_angle(dx, dy);
        pf.nx = dx / diag;
        pf.ny = dy / diag;
        pf.radius = own_radius / diag;
        pf.source = feat;
        positioned.push_back(std::move(pf));
    }
    return positioned;
}

// =============================================================================
// 2. Spell definitions
// =============================================================================

void RelativeDistanceConstraint::validate() const {
    if (relation != "farther" && relation != "closer" && relation != "inside" && relation != "outside") {
        throw std::invalid_argument(
            "RelativeDistanceConstraint.relation must be 'farther', 'closer', 'inside', or 'outside', got '" +
            relation + "'.");
    }
}

void SpellDefinition::validate() const {
    std::unordered_set<int> known_ids;
    for (const auto& slot : features) known_ids.insert(slot.id);
    for (const auto& c : relative_constraints) {
        c.validate();
        if (known_ids.find(c.subject_id) == known_ids.end()) {
            throw std::invalid_argument("Spell '" + name + "': RelativeDistanceConstraint.subject_id=" +
                                         std::to_string(c.subject_id) + " does not match any feature slot id.");
        }
        if (known_ids.find(c.reference_id) == known_ids.end()) {
            throw std::invalid_argument("Spell '" + name + "': RelativeDistanceConstraint.reference_id=" +
                                         std::to_string(c.reference_id) + " does not match any feature slot id.");
        }
    }
}

// =============================================================================
// 3. Matching
// =============================================================================

namespace {

// Score in [0.0, 1.0] for matching `feat` to `slot`, given they already
// share the same shape (shape match is a hard gate, checked by the caller
// via the by-shape candidate grouping in match_spell).
double slot_score(const SpellFeatureSlot& slot, const PositionedFeature& feat) {
    double dist_score;
    if (slot.distance.has_value()) {
        double dist_diff = std::abs(feat.distance - *slot.distance);
        if (dist_diff > slot.tolerance_dist) return 0.0;
        dist_score = 1.0 - (slot.tolerance_dist > 1e-9 ? (dist_diff / slot.tolerance_dist) : 0.0);
    } else {
        dist_score = 1.0;  // no fixed distance requirement
    }

    if (!slot.angle_constrained() || feat.distance <= config::SPELL_CENTER_EPSILON) {
        return dist_score;
    }

    if (angle_in_range(feat.angle, *slot.min_angle, *slot.max_angle)) {
        return dist_score;
    }

    double overshoot = sector_overshoot(feat.angle, *slot.min_angle, *slot.max_angle);
    double angle_score = std::max(0.0, 1.0 - overshoot / std::max(slot.tolerance_angle, 1e-9));
    if (angle_score <= 0.0) return 0.0;
    return dist_score * angle_score;
}

// Score in [0.0, 1.0] for one RelativeDistanceConstraint against a
// (possibly partial) slot assignment.
double relative_constraint_score(const RelativeDistanceConstraint& constraint,
                                  const std::unordered_map<int, const PositionedFeature*>& positioned_by_slot) {
    auto subj_it = positioned_by_slot.find(constraint.subject_id);
    auto ref_it = positioned_by_slot.find(constraint.reference_id);
    if (subj_it == positioned_by_slot.end() || ref_it == positioned_by_slot.end()) {
        return 0.0;
    }
    const PositionedFeature& subj = *subj_it->second;
    const PositionedFeature& ref = *ref_it->second;

    double tolerance = std::max(constraint.tolerance, 1e-9);
    double slack;

    if (constraint.relation == "farther") {
        slack = (subj.distance - ref.distance) - constraint.margin;
    } else if (constraint.relation == "closer") {
        slack = (ref.distance - subj.distance) - constraint.margin;
    } else if (constraint.relation == "inside") {
        double gap = std::hypot(subj.nx - ref.nx, subj.ny - ref.ny);
        slack = (ref.radius + constraint.margin) - gap;
    } else {  // "outside"
        double gap = std::hypot(subj.nx - ref.nx, subj.ny - ref.ny);
        slack = gap - (ref.radius + constraint.margin);
    }

    if (slack >= 0.0) return 1.0;
    return std::max(0.0, 1.0 + slack / tolerance);
}

}  // namespace

SpellMatchResult match_spell(const SpellDefinition& spell, const std::vector<std::shared_ptr<Feature>>& scene_features) {
    std::vector<PositionedFeature> positioned = compute_positions(scene_features);

    // Group candidates by shape up front.
    std::unordered_map<std::string, std::vector<int>> by_shape;
    for (size_t idx = 0; idx < positioned.size(); ++idx) {
        by_shape[positioned[idx].shape].push_back(static_cast<int>(idx));
    }

    const auto& slots = spell.features;
    const auto& constraints = spell.relative_constraints;
    int num_terms = static_cast<int>(slots.size() + constraints.size());

    std::vector<bool> used(positioned.size(), false);
    // current: slot_id -> positioned index. Using a std::map keeps
    // iteration order deterministic (not that it matters for correctness,
    // only for reproducibility).
    std::map<int, int> current;

    std::map<int, int> best;
    double best_score = -1.0;

    // Recursive backtracking search -- direct port of match_spell's
    // nested `backtrack` closure. `running_slot_score` accumulates the
    // sum of per-slot scores decided so far.
    std::function<void(size_t, double)> backtrack = [&](size_t slot_pos, double running_slot_score) {
        if (slot_pos == slots.size()) {
            double constraint_score = 0.0;
            if (!constraints.empty()) {
                std::unordered_map<int, const PositionedFeature*> positioned_by_slot;
                for (const auto& kv : current) positioned_by_slot[kv.first] = &positioned[static_cast<size_t>(kv.second)];
                for (const auto& c : constraints) {
                    constraint_score += relative_constraint_score(c, positioned_by_slot);
                }
            }
            double total = running_slot_score + constraint_score;
            double avg = num_terms ? (total / num_terms) : 0.0;
            if (avg > best_score) {
                best_score = avg;
                best = current;
            }
            return;
        }

        const SpellFeatureSlot& slot = slots[slot_pos];
        auto it = by_shape.find(slot.shape);
        if (it != by_shape.end()) {
            for (int idx : it->second) {
                if (used[static_cast<size_t>(idx)]) continue;
                double score = slot_score(slot, positioned[static_cast<size_t>(idx)]);
                if (score <= 0.0) continue;
                used[static_cast<size_t>(idx)] = true;
                current[slot.id] = idx;
                backtrack(slot_pos + 1, running_slot_score + score);
                current.erase(slot.id);
                used[static_cast<size_t>(idx)] = false;
            }
        }

        // Also try leaving this slot unfilled.
        backtrack(slot_pos + 1, running_slot_score);
    };

    backtrack(0, 0.0);

    SpellMatchResult out;
    std::vector<std::pair<int, PositionedFeature>> assignment;
    for (const auto& kv : best) {
        assignment.emplace_back(kv.first, positioned[static_cast<size_t>(kv.second)]);
    }
    bool all_filled = assignment.size() == slots.size();
    double final_score = std::max(best_score, 0.0);

    bool constraints_satisfied = true;
    if (!constraints.empty()) {
        std::unordered_map<int, const PositionedFeature*> positioned_by_slot;
        for (const auto& kv : best) positioned_by_slot[kv.first] = &positioned[static_cast<size_t>(kv.second)];
        for (const auto& c : constraints) {
            if (relative_constraint_score(c, positioned_by_slot) <= 0.0) {
                constraints_satisfied = false;
                break;
            }
        }
    }
    bool accepted = all_filled && constraints_satisfied;

    out.name = accepted ? std::optional<std::string>(spell.name) : std::nullopt;
    out.score = final_score;
    out.accepted = accepted;
    out.assignment = std::move(assignment);
    return out;
}

std::optional<SpellMatchResult> match_best_spell(const std::vector<SpellDefinition>& spells,
                                                   const std::vector<std::shared_ptr<Feature>>& scene_features) {
    std::optional<SpellMatchResult> best_result;
    for (const auto& spell : spells) {
        SpellMatchResult result = match_spell(spell, scene_features);
        if (result.accepted && (!best_result.has_value() || result.score > best_result->score)) {
            best_result = std::move(result);
        }
    }
    return best_result;
}

}  // namespace qrec

// -----------------------------------------------------------------------------
// Note on rotation invariance (see header docstring)
// -----------------------------------------------------------------------------
// To match a spell drawn at any rotation, angles would need to be measured
// relative to one designated "anchor" feature's bearing from center instead
// of true north -- i.e. angle = (compass_angle(dx, dy) - anchor_bearing) %
// 360 for every feature, anchor included (which then sits at angle 0 by
// definition). That's a change to compute_positions (needs to know which
// feature is the anchor) and to how SpellFeatureSlot::min_angle/max_angle
// are interpreted (relative bearing from the anchor, not compass bearing) --
// deliberately NOT done here since it trades away the ability to express
// "must be drawn upright", which most spells probably want.
//
// Note: relative-distance/containment constraints (RelativeDistanceConstraint)
// are already rotation-agnostic by construction -- "farther from center",
// "closer than", and "inside"/"outside" don't care about compass direction,
// only distance/radius -- so they keep working unchanged even if the above
// were ever implemented.